File: ruby/cyberapi.rb
require 'net/http'
require 'json'
require 'uri'

module CyberAPI
  # CyberAPI Ruby Client
  # 
  # Official Ruby client for the CyberAPI Threat Intelligence API
  #
  # @example Basic usage
  #   client = CyberAPI::Client.new(api_key: "your_api_key_here")
  #   result = client.check(domain: "example.com")
  #   puts result["risk_score"]
  class Client
    BASE_URL = "https://threats.cyberapi.io/api/v1"
    USER_AGENT = "CyberAPI-Ruby/1.0.0"
    
    attr_reader :api_key, :base_url, :timeout
    
    # Initialize a new CyberAPI client
    #
    # @param api_key [String] Your CyberAPI key (get one at https://threats.cyberapi.io)
    # @param base_url [String] API base URL (default: production)
    # @param timeout [Integer] Request timeout in seconds (default: 10)
    def initialize(api_key:, base_url: BASE_URL, timeout: 10)
      raise ArgumentError, "API key is required" if api_key.nil? || api_key.empty?
      
      @api_key = api_key
      @base_url = base_url.chomp('/')
      @timeout = timeout
    end
    
    # Check a domain or IP for threats
    #
    # @param domain [String, nil] Domain name to check (e.g., "example.com")
    # @param ip [String, nil] IP address to check (e.g., "1.2.3.4")
    # @return [Hash] Threat intelligence report
    # @raise [ArgumentError] If neither domain nor ip is provided
    # @raise [CyberAPI::Error] If API request fails
    #
    # @example Check a domain
    #   result = client.check(domain: "example.com")
    #   puts result["verdict"]
    #
    # @example Check an IP
    #   result = client.check(ip: "1.2.3.4")
    #   puts result["geo_location"]
    def check(domain: nil, ip: nil)
      raise ArgumentError, "Either domain or ip must be provided" if domain.nil? && ip.nil?
      
      params = {}
      params[:domain] = domain if domain
      params[:ip] = ip if ip
      
      uri = URI("#{@base_url}/check")
      uri.query = URI.encode_www_form(params)
      
      request = Net::HTTP::Get.new(uri)
      request['X-API-Key'] = @api_key
      request['Accept'] = 'application/json'
      request['User-Agent'] = USER_AGENT
      
      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: uri.scheme == 'https', read_timeout: @timeout) do |http|
        http.request(request)
      end
      
      handle_response(response)
    end
    
    # Quick check if domain/IP is malicious
    #
    # @param domain [String, nil] Domain name to check
    # @param ip [String, nil] IP address to check
    # @return [Boolean] True if risk_score >= 80
    #
    # @example
    #   if client.malicious?(domain: "suspicious-site.com")
    #     puts "Warning: Malicious domain detected!"
    #   end
    def malicious?(domain: nil, ip: nil)
      result = check(domain: domain, ip: ip)
      result['risk_score'].to_i >= 80
    end
    
    # Get list of threat signals detected
    #
    # @param domain [String, nil] Domain name to check
    # @param ip [String, nil] IP address to check
    # @return [Array<String>] List of signal strings
    #
    # @example
    #   signals = client.signals(domain: "example.com")
    #   puts "Detected signals: #{signals.join(', ')}"
    def signals(domain: nil, ip: nil)
      result = check(domain: domain, ip: ip)
      result['signals'] || []
    end
    
    private
    
    def handle_response(response)
      case response.code.to_i
      when 200..299
        JSON.parse(response.body)
      when 400..499
        error_data = JSON.parse(response.body) rescue {}
        message = error_data['detail'] || "HTTP #{response.code}"
        raise CyberAPI::ClientError.new(message, response.code.to_i)
      when 500..599
        error_data = JSON.parse(response.body) rescue {}
        message = error_data['detail'] || "HTTP #{response.code}"
        raise CyberAPI::ServerError.new(message, response.code.to_i)
      else
        raise CyberAPI::Error.new("Unexpected HTTP status: #{response.code}", response.code.to_i)
      end
    rescue JSON::ParserError => e
      raise CyberAPI::Error.new("Invalid JSON response: #{e.message}")
    end
  end
  
  # Base error class for CyberAPI
  class Error < StandardError
    attr_reader :status_code
    
    def initialize(message, status_code = nil)
      super(message)
      @status_code = status_code
    end
  end
  
  # Client error (4xx responses)
  class ClientError < Error; end
  
  # Server error (5xx responses)
  class ServerError < Error; end
end

File: python/cyberapi.py
"""
CyberAPI Python SDK
Official Python client for the CyberAPI Threat Intelligence API
"""

import requests
from typing import Optional, Dict, Any

class CyberAPI:
    """
    CyberAPI Client for Python
    
    Usage:
        >>> from cyberapi import CyberAPI
        >>> client = CyberAPI(api_key="your_api_key_here")
        >>> result = client.check(domain="example.com")
        >>> print(result['risk_score'])
    """
    
    def __init__(self, api_key: str, base_url: str = "https://threats.cyberapi.io/api/v1"):
        """
        Initialize CyberAPI client
        
        Args:
            api_key: Your CyberAPI key (get one at https://threats.cyberapi.io)
            base_url: API base URL (default: production)
        """
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update({
            'X-API-Key': api_key,
            'User-Agent': 'CyberAPI-Python-SDK/1.0.0'
        })
    
    def check(self, domain: Optional[str] = None, ip: Optional[str] = None) -> Dict[str, Any]:
        """
        Check a domain or IP for threats
        
        Args:
            domain: Domain name to check (e.g., "example.com")
            ip: IP address to check (e.g., "1.2.3.4")
            
        Returns:
            dict: Threat intelligence report
            
        Raises:
            ValueError: If neither domain nor ip is provided
            requests.HTTPError: If API request fails
        """
        if not domain and not ip:
            raise ValueError("Either 'domain' or 'ip' must be provided")
        
        params = {}
        if domain:
            params['domain'] = domain
        if ip:
            params['ip'] = ip
        
        response = self.session.get(f"{self.base_url}/check", params=params)
        response.raise_for_status()
        
        return response.json()
    
    def is_malicious(self, domain: Optional[str] = None, ip: Optional[str] = None) -> bool:
        """
        Quick check if domain/IP is malicious
        
        Returns:
            bool: True if risk_score >= 80
        """
        result = self.check(domain=domain, ip=ip)
        return result.get('risk_score', 0) >= 80
    
    def get_signals(self, domain: Optional[str] = None, ip: Optional[str] = None) -> list:
        """
        Get list of threat signals detected
        
        Returns:
            list: List of signal strings
        """
        result = self.check(domain=domain, ip=ip)
        return result.get('signals', [])
